File: frontend/src/lib/expenseReport.ts
import type { SavedInvoice, SavedLineItem } from '../types/invoice'

export const EXPENSE_REPORT_LOGO_PATH = '/expense-report-logo.png'

export type CategorySummaryRow = {
  category: string
  amount: number
}

const CATEGORY_CHART_COLORS = [
  '#334155',
  '#2563eb',
  '#0891b2',
  '#059669',
  '#d97706',
  '#dc2626',
  '#7c3aed',
  '#db2777',
] as const

export function getCategoryColor(index: number): string {
  return CATEGORY_CHART_COLORS[index % CATEGORY_CHART_COLORS.length]
}

export type ExpenseReportData = {
  vendor: string
  date: string
  invoiceNumber: string
  lineItems: Array<{
    index: number
    item: string
    category: string
    qty: string
    amount: number
  }>
  categorySummary: CategorySummaryRow[]
  subtotal: number
  tax: number | null
  total: number
}

export function formatReportAmount(value: number | null | undefined): string {
  if (value === null || value === undefined || !Number.isFinite(value)) {
    return '—'
  }
  return value.toFixed(2)
}

function displayField(value: string | null | undefined): string {
  const trimmed = value?.trim()
  return trimmed ? trimmed : '—'
}

function sumLineItems(lineItems: SavedLineItem[]): number {
  return lineItems.reduce((sum, item) => sum + (item.amount ?? 0), 0)
}

function buildCategorySummary(lineItems: SavedLineItem[]): CategorySummaryRow[] {
  const totals = new Map<string, number>()

  for (const item of lineItems) {
    const category = item.category?.trim() || 'Uncategorized'
    totals.set(category, (totals.get(category) ?? 0) + (item.amount ?? 0))
  }

  return Array.from(totals.entries())
    .map(([category, amount]) => ({ category, amount }))
    .sort((a, b) => a.category.localeCompare(b.category))
}

function resolveTaxAmount(invoice: SavedInvoice, subtotal: number): number | null {
  if (invoice.tax_amount != null) {
    return invoice.tax_amount
  }
  if (invoice.total_amount != null) {
    return invoice.total_amount - subtotal
  }
  return null
}

function resolveTotalAmount(invoice: SavedInvoice, subtotal: number, tax: number | null): number {
  if (invoice.total_amount != null) {
    return invoice.total_amount
  }
  if (tax != null) {
    return subtotal + tax
  }
  return subtotal
}

export function buildExpenseReportData(invoice: SavedInvoice): ExpenseReportData {
  const subtotal = invoice.subtotal_amount ?? sumLineItems(invoice.line_items)
  const tax = resolveTaxAmount(invoice, subtotal)
  const total = resolveTotalAmount(invoice, subtotal, tax)

  return {
    vendor: displayField(invoice.issuer.name),
    date: displayField(invoice.invoice_date),
    invoiceNumber: displayField(invoice.invoice_number),
    lineItems: invoice.line_items.map((item, index) => ({
      index: index + 1,
      item: displayField(item.description),
      category: displayField(item.category),
      qty: item.quantity === null ? '—' : String(item.quantity),
      amount: item.amount ?? 0,
    })),
    categorySummary: buildCategorySummary(invoice.line_items),
    subtotal,
    tax,
    total,
  }
}

File: backend/app/expense_report.py
from __future__ import annotations

from dataclasses import dataclass

from app.schemas import SavedInvoice, SavedLineItem

CATEGORY_CHART_COLORS = (
    "#334155",
    "#2563eb",
    "#0891b2",
    "#059669",
    "#d97706",
    "#dc2626",
    "#7c3aed",
    "#db2777",
)


def get_category_color(index: int) -> str:
    return CATEGORY_CHART_COLORS[index % len(CATEGORY_CHART_COLORS)]


@dataclass(frozen=True)
class CategorySummaryRow:
    category: str
    amount: float


@dataclass(frozen=True)
class ExpenseReportLineItem:
    index: int
    item: str
    category: str
    qty: str
    amount: float


@dataclass(frozen=True)
class ExpenseReportData:
    vendor: str
    date: str
    invoice_number: str
    line_items: list[ExpenseReportLineItem]
    category_summary: list[CategorySummaryRow]
    subtotal: float
    tax: float | None
    total: float


def format_report_amount(value: float | None) -> str:
    if value is None:
        return "—"
    return f"{value:.2f}"


def _display_field(value: str | None) -> str:
    trimmed = (value or "").strip()
    return trimmed if trimmed else "—"


def _sum_line_items(line_items: list[SavedLineItem]) -> float:
    return sum(item.amount or 0 for item in line_items)


def build_category_summary(line_items: list[SavedLineItem]) -> list[CategorySummaryRow]:
    totals: dict[str, float] = {}

    for item in line_items:
        category = (item.category or "").strip() or "Uncategorized"
        totals[category] = totals.get(category, 0) + (item.amount or 0)

    return sorted(
        (CategorySummaryRow(category=category, amount=amount) for category, amount in totals.items()),
        key=lambda row: row.category,
    )


def _resolve_tax_amount(invoice: SavedInvoice, subtotal: float) -> float | None:
    if invoice.tax_amount is not None:
        return invoice.tax_amount
    if invoice.total_amount is not None:
        return invoice.total_amount - subtotal
    return None


def _resolve_total_amount(invoice: SavedInvoice, subtotal: float, tax: float | None) -> float:
    if invoice.total_amount is not None:
        return invoice.total_amount
    if tax is not None:
        return subtotal + tax
    return subtotal


def build_expense_report_data(invoice: SavedInvoice) -> ExpenseReportData:
    subtotal = invoice.subtotal_amount if invoice.subtotal_amount is not None else _sum_line_items(invoice.line_items)
    tax = _resolve_tax_amount(invoice, subtotal)
    total = _resolve_total_amount(invoice, subtotal, tax)

    return ExpenseReportData(
        vendor=_display_field(invoice.issuer.name),
        date=_display_field(invoice.invoice_date),
        invoice_number=_display_field(invoice.invoice_number),
        line_items=[
            ExpenseReportLineItem(
                index=index + 1,
                item=_display_field(item.description),
                category=_display_field(item.category),
                qty="—" if item.quantity is None else str(item.quantity),
                amount=item.amount or 0,
            )
            for index, item in enumerate(invoice.line_items)
        ],
        category_summary=build_category_summary(invoice.line_items),
        subtotal=subtotal,
        tax=tax,
        total=total,
    )


def expense_report_filename(invoice: SavedInvoice) -> str:
    number = (invoice.invoice_number or "").strip()
    if number:
        safe = "".join(char if char.isalnum() or char in "._-" else "_" for char in number)
    else:
        safe = f"invoice-{invoice.id}"
    return f"expense-report-{safe}.pdf"
